### PR TITLE
feat(price): add support for onetime charges

### PR DESCRIPTION
**Pull request by omkar273**
### Problem Statement
Subscriptions could not carry **one-time (ONETIME)** fixed charges — charges that are invoiced exactly once at a specific date rather than recurring each billing cycle. The system had no way to represent or bill such charges, requiring workarounds like creating separate manual invoices outside the subscription lifecycle.

---

### Root Cause
The billing domain was built around the assumption that every line item maps 1-to-1 with a recurring billing cadence. Specifically:
- `ClassifyLineItems` only handled RECURRING items; no branch existed for ONETIME.
- `CalculateFixedCharges` always prorated charges over the billing period window, making it impossible to bill a flat amount on a single date.
- `FilterLineItemsToBeInvoiced` used a standard open-interval overlap check (`lineEnd > periodStart`) which is undefined when `lineStart == lineEnd` (zero-duration period).
- No DTO validation enforced constraints specific to ONETIME prices (e.g. `invoice_cadence` must be ADVANCE).
- The `SubscriptionLineItem` domain model stored a redundant `billing_cadence` column; the approach was simplified to derive ONETIME-ness from `BillingPeriod == ""` + `Price.BillingCadence`.

---

### Dev Approach

**Detection strategy** — `IsOneTime()` uses a two-level check:
1. Primary: `li.Price.BillingCadence == BILLING_CADENCE_ONETIME` (when price is pre-loaded).
2. Fallback: `li.PriceType == FIXED && li.BillingPeriod == ""` (for test fixtures / DB reads before price join).

`attachPricesToLineItems()` bulk-fetches prices before `ClassifyLineItems` so the primary path is always available in production.

**Classification** — ONETIME items use their `StartDate` as the single charge date:
- ADVANCE: classified when `chargeDate ∈ [currentPeriodStart, currentPeriodEnd)` (current) or `[nextPeriodStart, nextPeriodEnd)` (next). An item already in the current bucket is excluded from next.
- ARREAR: classified when `chargeDate ∈ (currentPeriodStart, currentPeriodEnd]`.

**Invoice generation** — `CalculateFixedCharges` skips proration for ONETIME items and emits a zero-duration period line `(PeriodStart = PeriodEnd = chargeDate)`.

**Deduplication** — `checkIfChargeInvoiced` uses a closed interval `[periodStart, periodEnd]` for zero-duration periods, matching the ARREAR end-inclusive boundary in `ClassifyLineItems`.

**Schema** — `billing_cadence` column was removed from `subscription_line_items` (V3 migration deleted); the information is derived at runtime from the price record.

**Phase 0 coupons** — `createPhaseExtraLineItems` return value is now captured; phase-level coupons are applied via `normalizePhaseCoupons` + `ApplyCouponsToSubscription` inside the transaction, before invoice generation.

---

### Flow

```
CreateSubscription / AddSubscriptionLineItem
         │
         ▼
  createPhaseExtraLineItems  ←── ONETIME line items stored with BillingPeriod=""
         │
         ▼
  CreateSubscriptionInvoice (ReferencePoint = PeriodStart)
         │
         ▼
  PrepareSubscriptionInvoiceRequest
    └─ attachPricesToLineItems()         ← bulk price fetch
    └─ ClassifyLineItems()
         ├─ RECURRING  → CurrentPeriodAdvance / Arrear / NextPeriodAdvance
         └─ ONETIME    → CurrentPeriodAdvance (ADVANCE) / CurrentPeriodArrear (ARREAR) / NextPeriodAdvance
         │
         ▼
  FilterLineItemsToBeInvoiced()
    └─ checkIfChargeInvoiced()  [closed interval for zero-duration]
         │
         ▼
  CalculateFixedCharges()
    └─ ONETIME → full amount, PeriodStart=PeriodEnd=chargeDate (no proration)
    └─ RECURRING → normal proration
         │
         ▼
  Invoice line items created
```

---

### Test Cases

| Test | Scenario | Expected |
|------|----------|----------|
| `TestClassify_OneTimeAdvance_ChargeDateAtPeriodStart` | chargeDate == periodStart | → CurrentPeriodAdvance |
| `TestClassify_OneTimeAdvance_ChargeDateMidPeriod` | chargeDate mid-period | → CurrentPeriodAdvance |
| `TestClassify_OneTimeAdvance_ChargeDateAtPeriodEnd` | chargeDate == periodEnd | → NextPeriodAdvance |
| `TestClassify_OneTimeArrear_ChargeDateAtPeriodEnd` | arrear, chargeDate == periodEnd | → CurrentPeriodArrear |
| `TestClassify_OneTimeArrear_ChargeDateAtPeriodStart` | arrear, chargeDate == periodStart | → excluded (exclusive lower bound) |
| `TestCalculateFixed_OneTime_FullAmountNoProration` | mid-month subscription start | Full $500, no proration |
| `TestCalculateFixed_OneTime_LineItemPeriodIsChargeDate` | PeriodStart/End in output | Both equal charge date |
| `TestOnetime_MultipleChargesInSamePeriod` | 2 ONETIME items same period | Both classified |
| `TestIsOneTime_ViaMapper` | ONETIME through ent→domain mapper | IsOneTime() == true |
| `TestIsOneTime_ViaMapper_Recurring` | RECURRING through mapper | IsOneTime() == false |

---

### Edge Cases to Be Aware Of

1. **Boundary at `periodEnd` (arrear)**: `chargeDate == periodEnd` is inclusive for ARREAR classification but was previously exclusive in `checkIfChargeInvoiced` (causing potential double-billing). Fixed with closed interval `[start, end]`.

2. **Zero-duration period overlap**: The standard overlap check `lineEnd > periodStart` evaluates to `false` when `lineStart == lineEnd == periodStart`. Handled by a dedicated zero-duration branch before the standard check.

3. **ONETIME in both current and next period**: An ONETIME item classified into `CurrentPeriodAdvance` is explicitly excluded from `NextPeriodAdvance` to prevent double classification.

4. **`IsActive()` for ONETIME items**: Without an explicit `EndDate`, a ONETIME item was "active forever" after its charge date. `IsActive()` now treats `EndDate == zero` as closed at `StartDate` for ONETIME items (used by usage-tracking path).

5. **`GetPeriodEnd()` for ONETIME items**: Returns `StartDate` when `EndDate` is zero, ensuring zero-duration window `[chargeDate, chargeDate]` rather than the full billing period end. Prevents mis-bucketing in any caller that uses this helper.

6. **ONETIME + `invoice_cadence` blank**: `CreatePriceRequest.Validate()` now normalises empty `invoice_cadence` to `ADVANCE` for ONETIME prices, preventing silent misclassification.

7. **Negative `billing_period_count` for ONETIME**: Now rejected with a validation error (`billing_period_count` is irrelevant for ONETIME but must not be negative).

8. **Phase 0 coupons**: Previously, coupons attached to phase 0 extra line items were silently ignored because `handleSubCoupons` runs before phase creation. Now resolved inside the transaction after `createPhaseExtraLineItems`.

9. **`start_date` required for ONETIME line items**: `createPhaseExtraLineItems` defaults `charge_date` to the phase start date when not provided. Callers must ensure this invariant is met before persisting.

10. **DB migration**: `billing_cadence` column was never merged to production — V3 migration was deleted and the ent schema was regenerated. Any dev env that applied V3 locally needs to run `ALTER TABLE subscription_line_items DROP COLUMN IF EXISTS billing_cadence;` manually.

----

## Summary by CodeRabbit

* **New Features**
  * Phase-level extra line items on subscription creation and per-line-item charge_date support
  * ONETIME prices require ADVANCE invoice cadence

* **Bug Fixes**
  * One-time charges billed at full amount (no proration) with correct invoice periods
  * Improved classification/invoicing and charge-date validation (included in opening invoice)
  * Support for zero-duration invoice periods for one-time charges

* **Tests**
  * Comprehensive tests covering one-time charge edge cases

* **Documentation**
  * Design and implementation plan for one-time charge cleanup

* **Chores**
  * DB migration to track billing cadence for line items

### CODE REVIEW

File: internal/api/dto/subscription_line_item.go [hunk 1]
```diff
@@ -94,6 +94,12 @@ type CreateSubscriptionLineItemRequest struct {
 	SubscriptionPhaseID  *string                         `json:"subscription_phase_id,omitempty"`
 	SkipEntitlementCheck bool                            `json:"-"` // This is used to skip entitlement check when creating a subscription line item
 
+	// ChargeDate is the date on which a one-time charge will be billed.
+	// Only applicable when the referenced price has billing_cadence = ONETIME.
+	// Defaults to the subscription (or phase) start date when not provided.
+	// Stored internally as the line item's start_date.
+	ChargeDate *time.Time `json:"charge_date,omitempty"`
+
```
Comment (nkmishra1997) [inline internal/api/dto/subscription_line_item.go]: Lets just get rid of this - use start_date for charge date logic in one time charges
add proper comments on each field so that user can understand the intended use of all fields of the DTO properly

File: internal/api/dto/subscription_line_item.go [hunk 2]
```diff
@@ -428,6 +467,20 @@ func (r *CreateSubscriptionLineItemRequest) validateCommitmentFields() error {
...
+		if params.Price.BillingCadence != "" {
+			billingCadence = params.Price.BillingCadence
+		}
+		invoiceCadence = params.Price.InvoiceCadence
+		// ONETIME charges default to ADVANCE invoice cadence if not explicitly set
+		if billingCadence == types.BILLING_CADENCE_ONETIME && invoiceCadence == "" {
+			invoiceCadence = types.InvoiceCadenceAdvance
+		}
+	}
+
```
Comment (nkmishra1997) [inline internal/api/dto/subscription_line_item.go]: why is this required
